---
title: "Seurat-DataViz"
author: "Elisabeth F. Heuston"
date: "2023-03-24"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "/Users/heustonef/Desktop/Obesity/scRNA/")
```

# Background

Ran `scRNASeq-Seurat.R` on lean, overweight, and obese HPAP scRNA data.  
lean (NW)  
overweight (OW)  
obese (OB)  
This document is primarily for data visualization. Additional analyses should be run in `scRNASeq-Seurat.R`  
# Setup
## Load libraries and local functions
```{r}
library(dplyr)
library(Seurat)
library(patchwork)
library(cowplot)
rnaProject <- "Obesity_scRNA"
writeLines(capture.output(sessionInfo()), paste0(rnaProject, "_sessionInfo.txt"))

```
## load local functions
```{r}
invisible(sapply(list.files("~/OneDrive-NIH/SingleCellMetaAnalysis/GitRepository/scMultiomics_MetaAnalysis/RFunctions/", 
														pattern = "*.R$", 
														full.names = TRUE), source))
```
## Load metadata
```{r}
metadata <- read.table(file = "/Users/heustonef/OneDrive-NIH/SingleCellMetaAnalysis/GitRepository/scMultiomics_MetaAnalysis/HPAPMetaData.txt", header = TRUE, sep = "\t")
rownames(metadata) <- metadata$DonorID
```
# Data Viz
## Load data
Note that when looking at differential gene expression, you should switch the asssay back to "SCT"
```{r}
seurat.object <- readRDS("Obesity_scRNA-AnchorIntegratedObject.RDS")
DefaultAssay(seurat.object) <- "SCT"
seurat.object$Obesity <- factor(seurat.object$Obesity, levels = c("NW", "OW", "OB"))

```
Some essential tables and info
```{r}
colnames(seurat.object@meta.data)
```
```{r}
table(seurat.object$DonorID)
```
```{r}
table(seurat.object$Obesity)
```

```{r}
table(seurat.object$Obesity, seurat.object$integrated_snn_res.0.5)
```
```{r}
table(seurat.object$integrated_snn_res.0.5, seurat.object$DonorID)
```

```{r}
counts.table <- table(seurat.object$Obesity, seurat.object$integrated_snn_res.0.5)
counts.table <- rbind(counts.table, colSums(counts.table))
counts.table <- as.data.frame.matrix(counts.table) %>%
	mutate(Tot.per.Class = rowSums(.))
print(counts.table)
```
Make percentages table
```{r}
pct.table <- data.frame(row.names = unique(seurat.object@meta.data$DonorID))
pct.table["tot.cls"] <- sapply(rownames(pct.table), FUN = function(x) nrow(seurat.object@meta.data[seurat.object@meta.data$DonorID == x,]))
pct.table["BMI"] <- sapply(rownames(pct.table), FUN = function(x) metadata[x, "BMI"])
for(i in levels(seurat.object@active.ident)){
	for(j in rownames(pct.table)){
		pct.table[j, paste0("integ_cl", i)] <- round((nrow(seurat.object@meta.data[seurat.object$DonorID == j & seurat.object$integrated_snn_res.0.5 == i,])/pct.table[j, "tot.cls"])*100, 2)
	}
}
# pct.table["DonorID"] <- rownames(pct.table)
pct.table
```

```{r}
p1 <- DimPlot(seurat.object, cols = color.palette, group.by = "Obesity")
p2 <- DimPlot(seurat.object, cols = color.palette, group.by = "integrated_snn_res.0.5")
p1 + p2

png(filename = paste0(rnaProject, "_UMAP-obesity.png"), height = 1200, width = 1200)
DimPlot(seurat.object, cols = color.palette, group.by = "Obesity")
dev.off()
png(filename = paste0(rnaProject, "_UMAP-intgsnnres05.png"), height = 1200, width = 1200)
DimPlot(seurat.object, cols = color.palette, group.by = "integrated_snn_res.0.5")
dev.off()

```
```{r}
DimPlot(seurat.object, cols = color.palette, group.by = "integrated_snn_res.0.5", split.by = "Obesity", ncol = 4)
```
```{r}
cM <- confusionMatrix(paste0(seurat.object$integrated_snn_res.0.5), paste0(seurat.object$Obesity))
cM <- cM / Matrix::rowSums(cM)
pheatmap::pheatmap(
	mat = as.matrix(cM),
	color = paletteContinuous("whiteBlue"),
	border_color = "black", 
	display_numbers = TRUE, 
	number_color = "firebrick"
)
png(filename = paste0(rnaProject, "-ClusterHeatmap.png"), height = 1000, width = 1200)
pheatmap::pheatmap(
	mat = as.matrix(cM),
	color = paletteContinuous("whiteBlue"),
	border_color = "black", 
	display_numbers = TRUE, 
	number_color = "firebrick"
)
dev.off()


```


# Load and mess with Panc markers
```{r}
panc.markers <- readRDS("~/OneDrive-NIH/SingleCellMetaAnalysis/GitRepository/scMultiomics_MetaAnalysis/RFunctions/panc.markers.RDS")
```


```{r}
VlnPlot(seurat.object, features = c("INS", "GCG", "PPY", "LEPR"), group.by = "integrated_snn_res.0.5", split.by = "Obesity", ncol = 1, pt.size = 0)
```
```{r}
FeaturePlot(seurat.object, features = c("INS", "GCG"), blend = TRUE)
```
```{r}
DotPlot(seurat.object, features = c("INS", "GCG", "PPY", "SST", "LEPR", "CPA1"), split.by = "Obesity", cols = c("gold", "blue3", "red3"))
```

```{r}
DotPlot(seurat.object, features = panc.markers$myeloid, split.by = "Obesity", cols = c("gold", "blue3", "red3"))
png(filename = "test.png", height = 1200, width = 800)
DotPlot(seurat.object, features = panc.markers$myeloid, split.by = "Obesity", cols = c("gold", "blue3", "red3"))
dev.off()
```



```{r}
#Need to figure out how to normalize this, since it's weighted by number of cells...
# how to plot % of cells
# pct.long <- reshape2::melt(pct.table[,-1], id.vars = "BMI")
# pct.long$BMI <- as.factor(pct.long$BMI)

VlnPlot(seurat.object, features = "BMI", group.by = "integrated_snn_res.0.5", cols = color.palette, pt.size = .1)
```

```{r}
# png(filename = paste0(rnaProject, "-SeuratDotplot-", i, ".png"), height = 1200, width = 800)
DotPlot(seurat.object, features = panc.markers[i], split.by = "Obesity", cols = c("gold", "blue3", "red3"), dot.min = )
# dev.off()

```

